clang-tools-extra/clangd/Preamble.cpp

```diff
@@ -587,11 +587,11 @@ class DiagPatcher {
 };
 } // namespace
 
-std::shared_ptr<const PreambleData>
-buildPreamble(PathRef FileName, CompilerInvocation CI,
-              const ParseInputs &Inputs, bool StoreInMemory,
-              PreambleParsedCallback PreambleCallback,
-              PreambleBuildStats *Stats) {
+std::shared_ptr<const PreambleData> buildPreamble(
+    PathRef FileName, CompilerInvocation CI, const ParseInputs &Inputs,
+    bool StoreInMemory, bool ExperimentalModulesSupport,
+    const GlobalCompilationDatabase &CDB,
```

Comment:
this dependency is surprising on this interface - contrast with ParseInputs which already contains the CompileCommand for the TU.

We need this to provide two capabilities:
 - in order to query the module graph: this can be more clearly represented by injecting `ProjectModules` here rather than the CDB. (And null can mean "no modules support").
 - in order to build the modules: in general this is going to become build-or-reuse once we add caching, so just the CDB doesn't provide enough state.

I think we should have a `ModuleBuilder` object which encapsulates build-or-reuse, deals with versioning of inputs etc.
The current `static PrerequisiteModules::buildPrerequisiteModulesFor()` would become a method on `ModuleBuilder`, and in future `ModuleBuilder` would own the cache.

WDYT? For now this is a pure refactoring: ModuleBuilder is just a wrapper around CDB that always builds. But I think `buildPreamble(..., ParseInputs, const ProjectModules*, const ModuleBuilder&)` shows responsibilities much more clearly than `buildPreamble(..., ParseInputs, bool ExperimentalModulesSupport, const GlobalCDB&)`